lib/wasi/src/runtime/module_cache/types.rs

```diff
@@ -0,0 +1,78 @@
+use std::{fmt::Debug, ops::Deref};
+
+use wasmer::{Engine, Module};
+
+use crate::runtime::module_cache::AndThen;
+
+/// A cache for compiled WebAssembly modules.
+///
+/// ## Assumptions
+///
+/// Implementations can assume that cache keys are unique and that using the
+/// same key to load or save will always result in the "same" module.
+///
+/// Implementations can also assume that [`CompiledModuleCache::load()`] will
+/// be called more often than [`CompiledModuleCache::save()`] and optimise
+/// their caching strategy accordingly.
+#[async_trait::async_trait]
+pub trait CompiledModuleCache: Debug + Send + Sync {
+    async fn load(&self, key: &str, engine: &Engine) -> Result<Module, CacheError>;
+
+    async fn save(&self, key: &str, module: &Module) -> Result<(), CacheError>;
```

Comment:
I just realized that we can make `load` and `save` perfectly sync. As I have not seen any await outside of calling load/save.

Which basically means that we can use wasmer-cache directly?

Reply:
No, there will very much be async use cases.

In a tokio context you will want to use `tokio::fs`, otherwise you need to spawn a blocking task.

In case of a shared/distributed cache we also need async.